### PR TITLE
[WIP][dashboard scoped filter] part 2: add algorithm to convert checked ids to scope object

### PR DESCRIPTION
**Pull request by graceguo-supercat**
this is #8521 rebased on latest master.

### REVIEWERS
@willbarrett @kristw @etr2460

### CODE REVIEW

File: superset/assets/src/dashboard/util/getFilterScopeFromNodesTree.js [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+  // aggregate scope to parentNodeValue
+  if (
+    forceAggregate ||
+    Object.entries(tabScopes).every(
+      ([key, { scope }]) => scope.length && key === scope[0],
+    )
+  ) {
+    return {
+      scope: [parentNodeValue],
+      immune: [].concat(
```
Comment (kristw) [inline superset/assets/src/dashboard/util/getFilterScopeFromNodesTree.js]: can use `flatMap`

```ts
import { flatMap } from 'lodash';

//...

immune: flatMap(Object.values(tabScopes), ({ immune }) => immune),

// ...

return {
  scope: flatMap(componentsInScope, ({ scope }) => scope), 
  immune: flatMap(componentsInScope, ({ immune }) => immune),
};
```

File: superset/assets/src/dashboard/util/getFilterScopeFromNodesTree.js [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+  filterKey,
+  nodes = [],
+  checkedChartIds = [],
+}) {
+  const { chartId } = getChartIdAndColumnFromFilterKey(filterKey);
+  if (nodes.length) {
+    return traverse({
+      currentNode: nodes[0],
+      filterId: chartId,
+      checkedChartIds,
```
Comment (kristw) [inline superset/assets/src/dashboard/util/getFilterScopeFromNodesTree.js]: can convert `checkedChartIds` to `Set`

File: superset/assets/src/dashboard/util/getFilterScopeFromNodesTree.js [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+  const { value: currentValue, children } = currentNode;
+  const chartChildren = children.filter(({ type }) => type === CHART_TYPE);
+  const tabChildren = children.filter(({ type }) => type === TAB_TYPE);
+
+  const chartsImmune = chartChildren
+    .filter(
+      ({ value }) => filterId !== value && !checkedChartIds.includes(value),
+    )
+    .map(({ value }) => value);
+  const tabScopes = tabChildren.reduce((map, child) => {
```
Comment (kristw) [inline superset/assets/src/dashboard/util/getFilterScopeFromNodesTree.js]: Can use `lodash/fp` for more functional programming style

```ts
import { flow, keyBy, mapValues } from 'lodash/fp';

const tabScopes = flow(
  keyBy(child => child.value),
  mapValues(child => traverse({
    currentNode: child,
    filterId,
    checkedChartIds,
  })),  
)(tabChildren);
```

File: superset/assets/src/dashboard/util/getFilterScopeFromNodesTree.js [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+  tabScopes,
+  parentNodeValue,
+  forceAggregate = false,
+}) {
+  // if all sub-tabs are in scope, or forceAggregate =  true
+  // aggregate scope to parentNodeValue
+  if (
+    forceAggregate ||
+    Object.entries(tabScopes).every(
+      ([key, { scope }]) => scope.length && key === scope[0],
```
Comment (kristw) [inline superset/assets/src/dashboard/util/getFilterScopeFromNodesTree.js]: `scope && scope.length` ?

File: superset/assets/src/dashboard/util/getFilterScopeFromNodesTree.js [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    immune: chartsImmune,
+  };
+}
+
+export default function getFilterScopeFromNodesTree({
+  filterKey,
+  nodes = [],
+  checkedChartIds = [],
+}) {
+  const { chartId } = getChartIdAndColumnFromFilterKey(filterKey);
```
Comment (kristw) [inline superset/assets/src/dashboard/util/getFilterScopeFromNodesTree.js]: Move this inside `if`

File: superset/assets/src/dashboard/util/getFilterScopeFromNodesTree.js [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+      forceAggregate: true,
+    });
+    return {
+      scope,
+      immune: chartsImmune.concat(immune),
+    };
+  }
+
+  // has tab children but only some sub-tab in scope
+  if (!isEmpty(tabChildren)) {
```
Comment (kristw) [inline superset/assets/src/dashboard/util/getFilterScopeFromNodesTree.js]: can avoid negate condition

File: superset/assets/src/dashboard/util/getFilterScopeFromNodesTree.js (hunk 2 above)
Comment (graceguo-supercat) [reply superset/assets/src/dashboard/util/getFilterScopeFromNodesTree.js]: `checkedChartIds` has no duplicates

**Review by kristw — APPROVED** [review]

**Pull request merged**